Add nbsphinx_timeout option

tbekolay opened the pull request:
This is the second PR. In our project, we have several cells that take quite a while to compute, so we encountered a `TimeoutError` when generating our docs.

I tried modifying the timeout through a config files in `~/.jupyter/jupyter_nbconvert_config.py`, but that does not appear to be read when import a preprocessor directory, so I had to add an option through `nbsphinx`. Not sure if there's another way to handle this? In any case, this works for us.

I also added a new notebook to the docs describing how to use the option, similar to the `allow_errors` option.

mgeier commented on nbsphinx.py:
```diff
@@ -386,6 +388,8 @@ class NotebookParser(rst.Parser):
 
     """
 
+    timeout = 30  # Default timeout
```
What does this do?
Is there some magic in the `rst.Parser` class?

tbekolay replied:
Oh whoops, this was leftover from an earlier attempt and I meant to take it out. Will amend!

tbekolay commented:
Amended!

mgeier commented on nbsphinx.py:
```diff
@@ -360,8 +361,9 @@ def from_notebook_node(self, nb, resources=None, **kw):
                 not any(c.outputs for c in nb.cells if 'outputs' in c)):
             allow_errors = (self._allow_errors or
                             nbsphinx_metadata.get('allow_errors', False))
+            timeout = nbsphinx_metadata.get('timeout', self._timeout)
```
Hmmm, should the `nbsphinx_timeout` value take precedence or the notebook metadata if both are given?

This line sure looks nicer than my 2 clumsy lines directly above, but they also have slightly different behavior, if I'm not mistaken ...

Either way, I think both `allow_errors` and `timeout` should behave in the same way, right?

tbekolay replied:
Yes, you're right, they are different. I'm happy to change, though we can't use the `or` idiom here since the default is an integer.

Perhaps the notebook metadata should take precedence? It feels to me like the notebook-local setting should take precedence over the global setting so that even when the global setting is in place notebooks can have specific behavior. What do you think?

mgeier replied:
I don't know.

I think cases can be made for both.

In your argumentation, `conf.py` sets "default" values.

However, one could also argue that those should "override" the notebook-local settings.

Either way, I think typically people will use either one or the other, but seldom both. So I guess in most cases it won't actually matter.

I think we should use your suggestion because the implementation is cleaner.
To make the behavior very clear, we should probably mention in the comments in `conf.py` that the settings can be overridden on a per-notebook basis.
We could also include those links:
http://nbsphinx.rtfd.org/en/latest/allow-errors.html
http://nbsphinx.rtfd.org/en/latest/timeout.html

If you want to, you can make those changes, if not, I'll make them once your PRs are merged.

tbekolay replied:
I agree on both comments: cases can be made for both, and people will tend to use one or the other. I'll update the behavior and update the `conf.py` and the notebooks :)

mgeier commented on doc/timeout.ipynb:
```diff
@@ -0,0 +1,60 @@
+    "\n",
+    "If you would like to change the amount of time given for a cell, you have three options:\n",
+    "\n",
+    "1. Manually execute the notebook in question and save the results, see [the pre-executed example notebook](pre-executed.ipynb#Exceptions).\n",
```
Could you please add a link in the other direction, too?

mgeier replied:
Also, I guess you wanted to link to `#Long-Running-Cells`, right?

tbekolay replied:
Yep, you're right. Will fix!

mgeier commented on doc/timeout.ipynb:
```diff
@@ -0,0 +1,60 @@
+ "cells": [
+  {
+   "cell_type": "markdown",
+   "metadata": {},
```
Can you please hide this cell with `"nbsphinx": "hidden"`?

mgeier commented on doc/conf.py:
```diff
@@ -16,6 +16,9 @@
 #nbsphinx_allow_errors = True
 
+# Controls when a cell will time out:
+#nbsphinx_timeout = 30  # Defaults to 30
```
I think I would prefer to mention the default value in the line before and use a different (non-default) value in the commented-out example (and no further comment on the commented-out line).

mgeier commented on doc/timeout.ipynb:
```diff
@@ -0,0 +1,60 @@
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# Long-running cells\n",
```
cells -> Cells?

I'm not a native speaker, so I'm not sure it that's proper, but that's what I used in all other section titles ...

tbekolay replied:
My apologies, I didn't look at the capitalization on the other titles. I'll update this to `Long-Running Cells` to be consistent!

mgeier replied:
I had a look at the other titles and I'm wondering how I should handle capitalization after hyphens ...

Currently, I have "Built-in" and "3rd-party" on the one side and "Pre-Executed", "Long-Running", "Sub-Directory" and "Sub-Section" on the other side.

Which do you think is better?

tbekolay replied:
Googling around a bit, there are a few style guides recommending various things... In one style, the test seems to be "would this word be understandable on its own?" and if so capitalize it. Another style says "no capital unless it would be capitalized in a normal sentence (e.g., proper nouns)."

The second style seems simpler to me, but I don't have a real preference, just to be consistent :) For now, I'll do `Long-Running` to be consistent, but we can change it later on.

tbekolay commented:
OK, I think I've addressed all your comments. I change the precedence for notebook-local and global settings in a separate commit (7243014002) to make that change more prominent.

mgeier commented:
Thanks a lot!
